File: android/core/protocol/src/main/kotlin/io/github/verybigsad/pimobile/protocol/Envelope.kt
package io.github.verybigsad.pimobile.protocol

import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive

private val typePattern = Regex("^[a-z][a-z0-9]*(?:[._-][a-z0-9]+)*$")

data class Envelope(
    val value: JsonObject,
) {
    val type: String get() = (value.getValue("type") as JsonPrimitive).content
}

fun parseEnvelope(text: String): Envelope {
    val value = parseJsonObject(text)
    val version = value["v"] as? JsonObject
    val major = (version?.get("major") as? JsonPrimitive)?.takeIf { !it.isString }?.content?.toDoubleOrNull()
    val minor = (version?.get("minor") as? JsonPrimitive)?.takeIf { !it.isString }?.content?.toDoubleOrNull()
    if (major != ProtocolConstants.major.toDouble() || minor == null || minor % 1.0 != 0.0 || minor !in 0.0..255.0) {
        throw ProtocolException(ProtocolErrorCode.UNSUPPORTED_VERSION, "Envelope version is unsupported")
    }
    val type = (value["type"] as? JsonPrimitive)?.takeIf { it.isString }?.content
    if (type == null || type.length !in 1..64 || !typePattern.matches(type)) protocolViolation("Envelope type is invalid")
    val messageId = (value["messageId"] as? JsonPrimitive)?.takeIf { it.isString }?.content
        ?: protocolViolation("Envelope messageId is invalid")
    uuidV4ToBytes(messageId)
    val replyTo = value["replyTo"]
    if (replyTo !is JsonPrimitive || (replyTo.content != "null" && !replyTo.isString)) protocolViolation("Envelope replyTo is invalid")
    if (replyTo.isString) uuidV4ToBytes(replyTo.content)
    if (value["body"] !is JsonObject) protocolViolation("Envelope body must be an object")
    return Envelope(value)
}

File: protocol/ts/src/envelope.ts
import { PROTOCOL_MAJOR, PROTOCOL_MINOR } from "./constants.js";
import { assertUuidV4 } from "./binary.js";
import { ProtocolError } from "./errors.js";
import type { JsonObject } from "./json.js";
import { assertJsonValue, isJsonObject } from "./json.js";

const TYPE_PATTERN = /^[a-z][a-z0-9]*(?:[._-][a-z0-9]+)*$/;

export interface Envelope extends JsonObject {
  readonly v: { readonly major: number; readonly minor: number };
  readonly type: string;
  readonly messageId: string;
  readonly replyTo: string | null;
  readonly body: JsonObject;
}

export function assertEnvelope(value: unknown): asserts value is Envelope {
  assertJsonValue(value);
  if (!isJsonObject(value)) {
    throw new ProtocolError("PROTOCOL_VIOLATION", "Envelope must be an object");
  }
  const version = value["v"];
  if (!isJsonObject(version) || version["major"] !== PROTOCOL_MAJOR || !Number.isInteger(version["minor"]) || (version["minor"] as number) < 0 || (version["minor"] as number) > 255) {
    throw new ProtocolError("UNSUPPORTED_VERSION", "Envelope version is unsupported");
  }
  const type = value["type"];
  if (typeof type !== "string" || type.length === 0 || type.length > 64 || !TYPE_PATTERN.test(type)) {
    throw new ProtocolError("PROTOCOL_VIOLATION", "Envelope type is invalid");
  }
  const messageId = value["messageId"];
  if (typeof messageId !== "string") {
    throw new ProtocolError("PROTOCOL_VIOLATION", "Envelope messageId is invalid");
  }
  assertUuidV4(messageId);
  const replyTo = value["replyTo"];
  if (replyTo !== null && typeof replyTo !== "string") {
    throw new ProtocolError("PROTOCOL_VIOLATION", "Envelope replyTo is invalid");
  }
  if (typeof replyTo === "string") {
    assertUuidV4(replyTo);
  }
  if (!isJsonObject(value["body"])) {
    throw new ProtocolError("PROTOCOL_VIOLATION", "Envelope body must be an object");
  }
}

export function createEnvelope(type: string, messageId: string, replyTo: string | null, body: JsonObject): Envelope {
  const value: JsonObject = { v: { major: PROTOCOL_MAJOR, minor: PROTOCOL_MINOR }, type, messageId, replyTo, body };
  assertEnvelope(value);
  return value;
}
